File: primalTestDotNet/Hero.cs
using PrimalTestDotNet.Entities;
using PrimalTestDotNet.Level;

namespace PrimalTestDotNet;

public class Hero(ICollider collider, int x, int y, IEnumerable<IGameObject> gameObjects, IDeathHandler deathHandler) : IDrawable
{
    public IntVector2 Position { get; set; } = new IntVector2(x, y);
    private IntVector2 LastPosition { get; set; } = new IntVector2(x, y);

    private int _health = 2;

    public int Health
    {
        get => _health;
        set
        {
            _health = Math.Clamp(value, 0, 2);
            if (_health == 0)
                deathHandler.OnDeath();
        }
    }

    public char Sprite => 'h';

    private readonly HashSet<IBackpackItem> _backpackItems = [];

    public bool HasSword => _backpackItems.Any(i => i is Sword);

    public bool HasTreasure => _backpackItems.Any(i => i is Treasure);

    public void AddItemToBackpack(IBackpackItem item)
    {
        _backpackItems.Add(item);
    }

    public void InteractWith(IGameObject interactable)
    {
        interactable.Interact(this);
    }

    public void Kill()
    {
        Health = 0;
    }

    public void Update()
    {
        // Iterating through a copy: this way the application wont crash on item removal
        foreach (var item in gameObjects.ToList())
        {
            if (item.Position.X == Position.X && item.Position.Y == Position.Y)
                item.Interact(this);
        }
    }

    public void Move(IntVector2 direction)
    {
        LastPosition = new IntVector2(Position.X, Position.Y);
        Position += direction;
        if (collider.IsCollidingWith(this))
            Position -= direction;
    }

    public void Draw(EntityRenderer renderer)
    {
        renderer.Draw(this);
    }

    public void JumpOverTrap()
    {
        var direction = Position - LastPosition;
        var nextPosition = Position + direction;

        if (collider.IsCollidingWith(nextPosition))
            Kill();
        else
            Move(direction);
    }
}

File: PrimalTestCpp/Hero.cpp
#include "Hero.h"
#include "EntityRenderer.h"
#include "ICollider.h"
#include "IDeathHandler.h"
#include "IBackpackItem.h"
#include "Sword.h"
#include "Treasure.h"
#include <vector>
#include <algorithm>

Hero::Hero(ICollider& collider, std::list<std::shared_ptr<IGameObject>>& gameObjects, int x, int y, IDeathHandler& deathHandler): 
	_position(x, y), _lastPosition(x, y), _collider(collider), _gameObjects(gameObjects), _deathHandler(deathHandler) {
	_health = 2;
}

void Hero::draw(const EntityRenderer& renderer) const
{
	renderer.draw(*this);
}

char Hero::getSprite() const
{
	return 'h';
}

const IntVector2D& Hero::getPosition() const
{
	return _position;
}

void Hero::jumpOverTrap() {
	auto direction = _position - _lastPosition;
	auto nextPosition = _position + direction;

	if (_collider.isCollidingWith(nextPosition))
		kill();
	else
		move(direction);
}

void Hero::move(const IntVector2D& direction) {
	_lastPosition = IntVector2D(_position.getX(), _position.getY());
	_position += direction;
	
	if (_collider.isCollidingWith(*this))
		_position -= direction;
}

void Hero::update() {
	std::vector<std::shared_ptr<IGameObject>> copy(_gameObjects.begin(), _gameObjects.end());
	for (auto& gameObject : copy) {
		if (gameObject->getPosition().getX() == _position.getX() && gameObject->getPosition().getY() == _position.getY())
			gameObject->interact(*this);
	}
}

void Hero::kill() {
	setHealth(0);
}

int Hero::getHealth() const {
	return _health;
}

void Hero::setHealth(int health) {
	_health = std::clamp(health, 0, 2);
	if (_health == 0)
		_deathHandler.onDeath();
}

bool Hero::hasSword() const {
	return std::any_of(_backpackItems.begin(), _backpackItems.end(), [](const auto& item) {
		return std::dynamic_pointer_cast<Sword>(item) != nullptr;
	});
}

bool Hero::hasTreasure() const {
	return std::any_of(_backpackItems.begin(), _backpackItems.end(), [](const auto& item) {
		return std::dynamic_pointer_cast<Treasure>(item) != nullptr;
		});
}

void Hero::addBackpackItem(std::shared_ptr<IBackpackItem> item) {
	_backpackItems.push_back(item);
}
